File: legacy/tools/wordingcheck.py
#!/usr/bin/env python3
"""Every dose claim in the tank summary must repeat the dosing engine's words.

The summary and the Dosing Wizard describe the same state. When each writes its
own sentence they drift, and the drift is invisible until someone reads both
screens: "parked off-target" on one while the other said "on its way to 475",
about the same element on the same day.

The rule is narrow on purpose. Claims about groups ("5 of 6 are in range") or
about stability rather than dosing are legitimately the summary's own voice.
What may not differ is a claim about a dose state, because the engine has
already worded that and the summary is repeating it.
"""
import re, sys

path = sys.argv[1] if len(sys.argv) > 1 else "src/reef-console.jsx"
src = open(path, encoding="utf-8").read()

m = re.search(r"\nfunction buildBriefing\(", src)
if not m:
    print("FAIL wordingcheck: buildBriefing not found")
    sys.exit(1)
start = m.start()
depth = 0
j = src.index("{", start)
k = j
while k < len(src):
    if src[k] == "{":
        depth += 1
    elif src[k] == "}":
        depth -= 1
        if depth == 0:
            break
    k += 1
body = src[start:k]

bad = []
checked = 0
for a in re.finditer(r"add\(\{[\s\S]{0,600}?\}\);", body):
    block = a.group(0)
    ident = re.search(r"id:\s*([^,\n]+)", block)
    claim = re.search(r"claim:\s*([^\n]+?),?\s*$", block, re.M)
    if not ident or not claim:
        continue
    ident_s, claim_s = ident.group(1).strip(), claim.group(1).strip().rstrip(",")
    if '"dose:"' not in ident_s:
        continue
    checked += 1
    # d.headline, or a ternary whose branches are all d.headline
    if not re.fullmatch(r"d\.headline", claim_s):
        bad.append((ident_s, claim_s[:70]))

if bad:
    print("FAIL dose claims that do not repeat the engine's wording:")
    for i, c in bad:
        print(f"    {i}: {c}")
    sys.exit(1)
print(f"OK   dose claims repeat the engine's wording ({checked} checked)")

File: scripts/verify/wordingcheck.mjs
#!/usr/bin/env node
/* Ported from legacy/tools/wordingcheck.py — unchanged logic, new path.
 *
 * Every dose claim in the tank summary must repeat the dosing engine's own
 * words. The summary and the Dosing Wizard describe the same state; when each
 * writes its own sentence they drift, invisibly, until someone reads both
 * screens on the same day. `buildBriefing` now lives in
 * src/lib/narrative-engine.js — this is the only automated enforcement of
 * wizard-states.md §7 and §12 that has ever existed.
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT, matchBracket } from './util.mjs';

const file = path.join(ROOT, 'src', 'lib', 'narrative-engine.js');
const src = fs.readFileSync(file, 'utf8');

const m = src.match(/\nexport function buildBriefing\(/);
if (!m) { console.log('FAIL wordingcheck: buildBriefing not found'); process.exit(1); }
const start = m.index;
const open = src.indexOf('{', start);
const close = matchBracket(src, open);
const body = src.slice(start, close + 1);

const bad = [];
let checked = 0;
for (const a of body.matchAll(/add\(\{[\s\S]{0,600}?\}\);/g)) {
  const block = a[0];
  const ident = block.match(/id:\s*([^,\n]+)/);
  const claim = block.match(/claim:\s*([^\n]+?),?\s*$/m);
  if (!ident || !claim) continue;
  const identS = ident[1].trim();
  const claimS = claim[1].trim().replace(/,$/, '');
  if (!identS.includes('"dose:"')) continue;
  checked++;
  if (!/^d\.headline$/.test(claimS)) bad.push([identS, claimS.slice(0, 70)]);
}

if (bad.length) {
  console.log('FAIL dose claims that do not repeat the engine\'s wording:');
  for (const [i, c] of bad) console.log(`    ${i}: ${c}`);
  process.exit(1);
}
console.log(`OK   dose claims repeat the engine's wording (${checked} checked)`);
